File: core/MajorityElementII/MajorityElementII.java
package com.dbc;

import java.util.ArrayList;
import java.util.List;

public class MajorityElementII {
    public List<Integer> majorityElement(int[] nums) {
        int num1 = 0, num2 = 0, vote1 = 0, vote2 = 0;
        for (int j : nums) {
            if (vote1 > 0 && num1 == j) vote1++;
            else if (vote2 > 0 && num2 == j) vote2++;
            else if (vote1 == 0) {
                num1 = j;
                vote1++;
            } else if (vote2 == 0) {
                num2 = j;
                vote2++;
            } else {
                vote1--;
                vote2--;
            }
        }

        int count1 = 0, count2 = 0;
        for (int num : nums) {
            if (vote1 > 0 && num == num1) count1++;
            if (vote2 > 0 && num == num2) count2++;
        }
        List<Integer> res = new ArrayList<>();
        if (count1 > nums.length / 3) res.add(num1);
        if (count2 > nums.length / 3) res.add(num2);
        return res;
    }
}

File: core/MajorityElementII/MajorityElementII.py
from typing import List


def majorityElement(nums: List[int]) -> List[int]:
    num1, num2, count1, count2 = 0, 0, 0, 0
    for i in range(len(nums)):
        if count1 > 0 and num1 == nums[i]:
            count1 += 1
        elif count2 > 0 and num2 == nums[i]:
            count2 += 1
        elif count1 == 0:
            num1 = nums[i]
            count1 = 1
        elif count2 == 0:
            num2 = nums[i]
            count2 = 1
        else:
            count1 -= 1
            count2 -= 1

    vote1, vote2 = 0, 0
    for num in nums:
        if count1 > 0 and num == num1:
            vote1 += 1
        if count2 > 0 and num == num2:
            vote2 += 1

    res = []
    if count1 > 0 and vote1 > len(nums) / 3:
        res.append(num1)
    if count2 > 0 and vote2 > len(nums) / 3:
        res.append(num2)

    return res




